### PR TITLE
Add curvature measurements

### PR DESCRIPTION
**Pull request by jo-mueller**
I added functions to measure surface curvature with a few different approaches:
* Built in vedo curvature measurements
* Custom sphere-fit curvature. This returns two surface layers with curvature and fit residues, resppectively, as values.

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -37,4 +37,6 @@ deps =
     pytest-qt
     qtpy
     pyqt5
+	vedo
+	open3d
```
Comment (haesleinhuepf) [inline tox.ini:41]:
```suggestion
```
I think those are not necessary as they are specified in setup.cfg

File: setup.cfg [hunk 2]
```diff
@@ -1,7 +1,7 @@
 [metadata]
 name = napari-process-points-and-surfaces
-version = 0.1.2
-author = Robert Haase
+version = 0.1.3
```
Comment (haesleinhuepf) [inline setup.cfg:3]:
```suggestion
version = 0.1.2
```
I'd like to bump the version when a new release is deployed

File: napari_process_points_and_surfaces/_quantification.py [hunk 3]
```diff
@@ -58,4 +66,55 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
     faces = np.asarray(mesh2.faces())
     values = np.asarray(mesh2.pointdata[mesh2.pointdata.keys()[0]])
 
-    return (vertices, faces, values)
\ No newline at end of file
+    return (vertices, faces, values)
+
+
+@register_function(menu="Measurement > Surface curvature (vedo, nppas)")
+def add_curvature(surface: SurfaceData, curvature_id: Curvature = Curvature.Gauss_Curvature) -> SurfaceData:
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:73]: How about naming the function "add_curvature_scalars"?

And could you please add a link to the vedo documentation and a docstring?

https://vedo.embl.es/autodocs/content/vedo/mesh.html#vedo.mesh.Mesh.addCurvatureScalars

File: napari_process_points_and_surfaces/_quantification.py [hunk 4]
```diff
@@ -58,4 +66,55 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
...
+
+    mesh = vedo.mesh.Mesh((surface[0], surface[1]))    
+    mesh.addCurvatureScalars(method=curvature_id.value)
+    
+    values = mesh.pointdata[curvature_id.name]
+    
+    return (mesh.points(), np.asarray(mesh.faces()), values)
+
+@register_function(menu="Measurement > Sphere-fitted surface curvature (nppas)")
+def spherefitted_curvature(surface: SurfaceData, radius: float = 1.0) -> List[LayerDataTuple]:
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:84]: How about naming the function "measure_sphere_fitted_curvature"?

File: napari_process_points_and_surfaces/__init__.py [hunk 5]
```diff
@@ -1,5 +1,5 @@
 
-__version__ = "0.1.2"
+__version__ = "0.1.3"
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/__init__.py:2]:
```suggestion
__version__ = "0.1.2"
```
Leave version bumps to the maintainer ;-)

File: napari_process_points_and_surfaces/_quantification.py [hunk 6]
```diff
@@ -40,6 +42,12 @@ class Quality(Enum):
     ASPECT_GAMMA = 27
     AREA = 28
     ASPECT_BETA = 29
+    
+class Curvature(Enum):
+    Gauss_Curvature = 0
+    Mean_Curvature = 1
+    Maximum_Curvature = 2
+    Minimum_Curvature = 3
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:50]: According to the vedo documentation, those are called:
0-gaussian, 1-mean, 2-max, 3-min curvature

Can we please have the same names here? This makes it later easier to use.

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hi Johannes,

super cool, thanks for your contribution here! I'd like to ask for some changes. In the other functions I tried to mimic the vedo and open3d API. Hence, I'd suggest to name a function "add_curvature_scalars" if a function with this name exists in vedo and we're just doing the same here. Also some docstrings with explanations of what functions do would be great.

Otherwise great! I will merge soon after you did these little modifications.

Thanks again :-)

File: napari_process_points_and_surfaces/_quantification.py [hunk 7]
```diff
@@ -58,4 +66,55 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
...
+        msgBox.setWindowTitle("Warning")
+        msgBox.setText(f"The chosen curvature radius ({radius})"
+                       "was too small to calculate curvatures. Increase " 
+                       "the radius to silence this error.")
+        msgBox.addButton(QPushButton('Ok'), QMessageBox.YesRole)
+        msgBox.exec()
+        return 0
+        
+    props_curv = {'name': 'curvature', 'colormap': 'viridis'}
+    props_res = {'name': 'fit residues', 'colormap': 'magma'}
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:115]: One more thing: If a variable contains "probability fit residues", please call the variable "probability_fit_residues". "pro_res" is no meaningful name. Also consider renaming "props_curv". Please use english words for variable names. Thank you!

File: napari_process_points_and_surfaces/_quantification.py (hunk 6 above)
Comment (jo-mueller) [reply napari_process_points_and_surfaces/_quantification.py:50]: Yes and no - they're only half called that way. They are referred as such inn the documentation but if you wanted to retrieve the actual data from the pointcloud object you would have to use `mesh.pointdata['Gauss_Curvature ']` or `mesh.pointdata['Maximum_Curvature ']`. I find the API not 100% consistent on this one.

Comment (haesleinhuepf) [reply napari_process_points_and_surfaces/_quantification.py:50]: 👍

File: tox.ini (hunk 1 above)
Comment (jo-mueller) [reply tox.ini:41]: Really? I didn't know, cool :)

**Comment by jo-mueller** [comment]
Hey @haesleinhuepf 

thanks for the fast feedback! Next time feel free to wait a little longer until I change the PR to be ready for review :)

**Comment by jo-mueller** [comment]
PS: I ran tests with tox which run for Python 3.8 and 3.9 but fail for 3.7, which seems to be a pytest issue rather than an issue with the tests/the functions

File: napari_process_points_and_surfaces/__init__.py [hunk 8]
```diff
@@ -54,6 +62,10 @@ def example_data_knot(viewer:napari.viewer):
 def example_data_standford_bunny(viewer:napari.viewer):
     viewer.add_surface(_standford_bunny(), blending='additive', shading='smooth')
 
+@register_action(menu = "Surfaces > Example data: Ellipsoid (vedo)")
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/__init__.py:65]:
```suggestion
@register_action(menu = "Surfaces > Example data: Ellipsoid (vedo,)")
```

File: napari_process_points_and_surfaces/_quantification.py [hunk 9]
```diff
@@ -58,4 +66,110 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
...
+    ----------
+    surface : SurfaceData
+        3-Tuple of (points, faces, values)
+    curvature_id : Union[Curvature, int] optional
+        Method to be used: 0-gaussian, 1-mean, 2-max, 3-min curvature. The
+        default is 0 (gaussian).
+    Returns
+    -------
+    SurfaceData
+        3-tuple consiting of (points, faces, values)
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:92]:
```suggestion
        3-tuple consisting of (points, faces, values)
```

File: napari_process_points_and_surfaces/_quantification.py [hunk 10]
```diff
@@ -58,4 +66,110 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
...
+    radius : float, optional
+        Radius within which other points of the surface will be considered
+        neighbors. The default is 1.0.
+
+    Returns
+    -------
+    List[LayerDataTuple]
+        A list of surface data items. The first entry corresponds to the curvature-
+        anotated surface, the second entry corresponds to the fit
+        residue-annotated surface.
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:133]: Can you maybe add a sentence above that explains what a "fit residue-annotated surface" is?

File: napari_process_points_and_surfaces/_quantification.py [hunk 11]
```diff
@@ -58,4 +66,110 @@ def add_quality(surface: SurfaceData, quality_id: Quality = Quality.MIN_ANGLE) -
...
+        msgBox.exec()
+        return 0
+        
+    properties_curvature_layer = {'name': 'curvature', 'colormap': 'viridis'}
+    properties_residues_layer = {'name': 'fit residues', 'colormap': 'magma'}
+        
+    layer1 = ((mesh.points(), np.asarray(mesh.faces()), curvature), properties_curvature_layer, 'surface')
+    layer2 = ((mesh.points(), np.asarray(mesh.faces()), residues), properties_residues_layer, 'surface')
+        
+    return [layer1, layer2]
```
Comment (haesleinhuepf) [inline napari_process_points_and_surfaces/_quantification.py:175]:
```suggestion
    return [layer1, layer2]

```

**Review by haesleinhuepf — APPROVED** [review]
Hi Johannes @jo-mueller ,

this all looks good to me, just some minor things.

Thanks!
Robert

File: napari_process_points_and_surfaces/_quantification.py (hunk 10 above)
Comment (jo-mueller) [reply napari_process_points_and_surfaces/_quantification.py]:
```
    List[LayerDataTuple]
        A list of surface data items. The items correspond to the curvature- and
        fit residue-annotated surface, respectively. 
        With each item consisting of a `(points, faces, values)` tuple, the 
        `value` variable reflects each vertice's curvature or fit residue result.'
```

**Comment by haesleinhuepf** [comment]
Great work Johannes @jo-mueller! Thanks a lot!

**Pull request merged**